### PR TITLE
Siamese comparison new

### PR DESCRIPTION
**Pull request by madisonmay**

### CODE REVIEW

**Comment by madisonmay** [comment]
As a heads up I've just noticed that the link to the quora dataset has died.  We'll be putting up a new file on s3 here soon if you want to use that for experimentation @chaitjo.  Will mention you in that PR so you know when that's merged in.

**Comment by madisonmay** [comment]
Your branch compiled and started training on the Quora dataset, so things are looking good!  You'll have to rebase onto the `origin/development` branch of finetune to get the proper link to the quora dataset, and figure out the details of the negative sampling / batching still, but otherwise looking good!

Depending on what hardware you're running on  and what batch size you're using you may also want to use `low_memory_mode=True`.  This uses an optimization that prevents finetune from having to store all activations in memory at once, and reduces the amount of GPU memory required for training at a slight performance hit. 

```
model = Comparison(verbose=True, n_epochs=1, low_memory_mode=True, batch_size=8)
```

File: finetune/siamese_comparison.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        kwargs['mask'] = np.stack([arr_0.mask, arr_1.mask], 1)
+        return ArrayEncodedOutput(**kwargs)
+
+    def _define_placeholders(self, target_dim=None):
+        super()._define_placeholders(target_dim=target_dim)
+        self.X = tf.placeholder(tf.int32, [None, 2, self.config.max_length, 2])
+        self.M = tf.placeholder(tf.float32, [None, 2, self.config.max_length])  # sequence mask
+
+    def _target_model(self, *, featurizer_state, targets, n_outputs, train=False, reuse=None, **kwargs):
+        hidden_0, hidden_1 = tf.split(featurizer_state["features"], num_or_size_splits=2, axis=1)
```
Comment (benleetownsend) [inline finetune/siamese_comparison.py]: These three lines can be achieved with one tf.unstack

File: finetune/network_modules.py [hunk 2]
```diff
@@ -312,3 +312,37 @@ def seq_lab_internal(hidden):
                 'transition_matrix': transition_params
             }
         }
+
+def cosine_similarity(hidden_0, hidden_1, targets, n_targets, dropout_placeholder, config, train=False, reuse=None, **kwargs):
```
Comment (benleetownsend) [inline finetune/network_modules.py]: Can somebody explain this too me, my understanding is that you'd have something that would look like. 
```
cos_sim_logits = tf.reduce_sum(tf.multiply(hidden1, hidden2))
loss = tf.nn.sigmoid_cross_entropy_with_logits(labels=targets, logits=cos_sim_logits)
```
where targets shape is [batch, 1]

**Comment by benleetownsend** [comment]
Note that the Quora dataset already has something that looks like hard-negative sampling. Where they have true labels given to true matches, and false labels given to samples that under a simple model might look the same but aren't. I'm working on refactoring the input code now that should make it easier to do a more standard Siamese network sampling but for the meanwhile, you can do an offline sampling method similar to what has been done for Quora.

**Comment by madisonmay** [comment]
@chaitjo -- I stand corrected.  I accidentally ran our Comparison model instead of your new `SiameseComparison` model.  Your branch doesn't yet compile and start training.  Taking a look at the errors raised now.

**Comment by madisonmay** [comment]
Superceded by #147
